web/app/js/components/Top.jsx

```diff
@@ -0,0 +1,361 @@
+        toNamespace: "",
+        method: "",
+        path: "",
+        scheme: "",
+        authority: "",
+        maxRps: defaultMaxRps
+      },
+      maxRowsToStore: 40,
+      awaitingWebSocketConnection: false,
+      tapRequestInProgress: false,
+      pollingInterval: 10000,
+      pendingRequests: false
+    };
+  }
+
+  componentDidMount() {
+    this.startServerPolling();
+  }
+
+  componentWillUnmount() {
+    if (this.ws) {
+      this.ws.close(1000);
+    }
+    this.stopTapStreaming();
+    this.stopServerPolling();
+  }
+
+  onWebsocketOpen = () => {
+    let query = _.cloneDeep(this.state.query);
+    query.maxRps = parseFloat(query.maxRps);
+
+    this.ws.send(JSON.stringify({
+      id: "tap-web",
+      ...query
+    }));
+    this.setState({
+      awaitingWebSocketConnection: false,
+      error: null
+    });
+  }
+
+  onWebsocketRecv = e => {
+    this.indexTapResult(e.data);
+  }
+
+  onWebsocketClose = e => {
+    this.stopTapStreaming();
+
+    if (!e.wasClean) {
+      this.setState({
+        error: {
+          error: `Websocket [${e.code}] ${e.reason}`
+        }
+      });
+    }
+  }
+
+  onWebsocketError = e => {
+    this.setState({
+      error: { error: e.message }
+    });
+
+    this.stopTapStreaming();
+  }
+
+  getResourcesByNs(rsp) {
+    let statTables = _.get(rsp, [0, "ok", "statTables"]);
+    let authoritiesByNs = {};
+    let resourcesByNs = _.reduce(statTables, (mem, table) => {
+      _.map(table.podGroup.rows, row => {
+        if (!mem[row.resource.namespace]) {
+          mem[row.resource.namespace] = [];
+          authoritiesByNs[row.resource.namespace] = [];
+        }
+
+        switch (row.resource.type.toLowerCase()) {
+          case "service":
+            break;
+          case "authority":
+            authoritiesByNs[row.resource.namespace].push(row.resource.name);
+            break;
+          default:
+            mem[row.resource.namespace].push(`${row.resource.type}/${row.resource.name}`);
+        }
+      });
+      return mem;
+    }, {});
+    return {
+      authoritiesByNs,
+      resourcesByNs
+    };
+  }
+
+  parseTapResult = data => {
+    let d = processTapEvent(data);
+
+    if (d.eventType === "rsp") {
+      d.success = parseInt(d.http.responseInit.httpStatus, 10) < 500;
+    } else if (d.eventType === "end") {
+      d.latency = parseFloat(d.http.responseEnd.sinceRequestInit.replace("s", ""));
+      d.completed = true;
+    }
+
+    return d;
+  }
+
+  topEventKey = event => {
+    return [event.source.str, event.destination.str, event.http.requestInit.path].join("_");
+  }
+
+  initialTopResult(d, eventKey) {
+    return {
+      count: 1,
+      best: d.end.latency,
+      worst: d.end.latency,
+      last: d.end.latency,
+      success: !d.rsp.success ? 0 : 1,
+      failure: !d.rsp.success ? 1 : 0,
+      successRate: !d.rsp.success ? new Percentage(0, 1) : new Percentage(1, 1),
+      source: d.req.source.str,
+      destination: d.req.destination.str,
```

Comment:
It would be really sweet if we could use the same Popover display for source and dest that we're using in the tap table:

![screen shot 2018-08-14 at 5 53 21 pm](https://user-images.githubusercontent.com/9226/44127318-a7cefd60-9ff1-11e8-8c7d-2adeb9937e0d.png)

The IPs in this table are a bit less exciting:

![screen shot 2018-08-14 at 6 37 02 pm](https://user-images.githubusercontent.com/9226/44127338-bdfd5e2e-9ff1-11e8-8084-39a0acef27d7.png)

Note that this relates to my other comment about aggregating tap events and can totally be done in a follow-up branch.